cpp/src/parquet/arrow/schema.cc

```diff
@@ -1053,6 +1060,29 @@ Result<bool> ApplyOriginalMetadata(const Field& origin_field, SchemaField* infer
       // inferred_type is arrow::extension::json(arrow::utf8())
       auto origin_storage_field = origin_field.WithType(ex_type.storage_type());
 
+      // Apply metadata recursively to storage type
+      RETURN_NOT_OK(ApplyOriginalStorageMetadata(*origin_storage_field, inferred));
+      inferred->field = inferred->field->WithType(origin_type);
+    } else if (inferred_type->id() == ::arrow::Type::FIXED_SIZE_BINARY &&
```

Comment:
These branches are growing longer while they look pretty similar. Not sure if we can refactor it a little bit to look nicer.